In [1]:
from selenium import webdriver
from webdriver_manager.chrome import ChromeDriverManager
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.common.exceptions import TimeoutException

In [52]:
driver = webdriver.Chrome(ChromeDriverManager().install())
wait = WebDriverWait(driver, 30)

[WDM] - 

[WDM] - ====== WebDriver manager ======
[WDM] - Current google-chrome version is 95.0.4638
[WDM] - Get LATEST driver version for 95.0.4638
[WDM] - Driver [C:\Users\YH\.wdm\drivers\chromedriver\win32\95.0.4638.54\chromedriver.exe] found in cache
<ipython-input-52-f004ae8d437d>:1: DeprecationWarning: executable_path has been deprecated, please pass in a Service object
  driver = webdriver.Chrome(ChromeDriverManager().install())


In [53]:
def search():
    try:
        driver.get('https://detail.tmall.com/item.htm?spm=a230r.1.14.161.3446631fe8NHUb&id=653068620195&ns=1&abbucket=7')
        # 等待搜索框加载出来
    except TimeoutException:
        # 如果出现超时异常，重新执行search()函数
        return search()

In [8]:
def next_page(i):
    try:
        if i < 7:
            page_selector = '#J_Reviews > div > div.rate-page > div > a:nth-child({})'.format(i+5)
            submit = wait.until(
                EC.element_to_be_clickable((By.CSS_SELECTOR, page_selector))
            )
            # 点击确定，进行翻页
            submit.click()
        elif i == 26:
            page_selector = '#J_Reviews > div > div.rate-page > div > a:nth-child({})'.format(9)
            submit = wait.until(
                EC.element_to_be_clickable((By.CSS_SELECTOR, page_selector))
            )
            # 点击确定，进行翻页
            submit.click()
        else:
            page_selector = '#J_Reviews > div > div.rate-page > div > a:nth-child({})'.format(11)
            submit = wait.until(
                EC.element_to_be_clickable((By.CSS_SELECTOR, page_selector))
            )
            # 点击确定，进行翻页
            submit.click()
    except TimeoutException:
        next_page(i)

In [9]:
# import PyQuery to strip html to text
from pyquery import PyQuery as pq

In [50]:
global_reviews = []

def get_products():
    # 确保商品信息完全加载完毕
    wait.until(
        EC.presence_of_element_located((By.CSS_SELECTOR, '#J_Reviews .tm-rate .tm-rate-content'))
    )
    # 获取网页源码
    html = driver.page_source
    doc = pq(html)
    ## 注意这个地方需要在下面的item find（）中的div的上一层div！！！只用管div，tr之类的都不用管
    items1 = doc('#J_Reviews .tm-rate .tm-rate-content').items()
    #items2 = doc('#J_Reviews .tm-rate .rate-grid').items()
    #items3 = doc('#J_Reviews .tm-rate .rate-grid').items()
    reviews = []
    #titles = []
    #dates = []
    # 获取评论
    for item in items1:
        reviews.append(item.find('.tm-rate-fulltxt').text())
    # 获取颜色和尺码    
    #for item in items2:
    #    titles.append(item.find('.rate-sku').text())
    # 获取日期
    #for item in items3:
    #    dates.append(item.find('.tm-rate-date').text())
    
    #把颜色尺码 和 日期list进行split
    #titles = titles[0].split(' ')
    #dates = dates[0].split(' ')
    
    # drop '此用户没有填写评论!'的情况，因为会多一个element，之后没有办法和时间对应
    #while '此用户没有填写评论!' in reviews:
    #    reviews.remove('此用户没有填写评论!')
    #    dates.remove('')
    
    for i in range(len(reviews)):
        global_reviews.append(reviews[i])
    #    global_titles.append(titles[i])
    #    global_dates.append(dates[i])
    
    print('Successfully scraped!')

In [35]:
def hold_slider():
    try:
        slider = driver.find_element_by_xpath("/html/body/div/div[2]/div/div[1]/div[2]/center/div[1]/div/div[1]/span")
        if slider.is_displayed():
            ActionChains(driver).click_and_hold(on_element=slider).perform()
            ActionChains(driver).move_by_offset(xoffset=258, yoffset=0).perform()
            ActionChains(driver).pause(0.5).release().perform()
    except:
        pass

In [64]:
len(global_reviews)

1403

In [60]:
def main():
    for i in range(50,100):
        get_products()
        next_page(i)

if __name__ == "__main__":
    main()

Successfully scraped!
Successfully scraped!
Successfully scraped!
Successfully scraped!
Successfully scraped!
Successfully scraped!
Successfully scraped!
Successfully scraped!
Successfully scraped!
Successfully scraped!
Successfully scraped!
Successfully scraped!
Successfully scraped!
Successfully scraped!
Successfully scraped!
Successfully scraped!
Successfully scraped!
Successfully scraped!
Successfully scraped!


StaleElementReferenceException: Message: stale element reference: element is not attached to the page document
  (Session info: chrome=95.0.4638.69)
Stacktrace:
Backtrace:
	Ordinal0 [0x00B306F3+2492147]
	Ordinal0 [0x00AC9BD1+2071505]
	Ordinal0 [0x009D2478+1057912]
	Ordinal0 [0x009D4BA4+1067940]
	Ordinal0 [0x009D4A6E+1067630]
	Ordinal0 [0x009D4CD0+1068240]
	Ordinal0 [0x009FE0FC+1237244]
	Ordinal0 [0x009F3573+1193331]
	Ordinal0 [0x00A157C3+1333187]
	Ordinal0 [0x009F3466+1193062]
	Ordinal0 [0x00A158AA+1333418]
	Ordinal0 [0x00A24F38+1396536]
	Ordinal0 [0x00A1568B+1332875]
	Ordinal0 [0x009F21D4+1188308]
	Ordinal0 [0x009F302F+1191983]
	GetHandleVerifier [0x00CB67A6+1545030]
	GetHandleVerifier [0x00D6105C+2243580]
	GetHandleVerifier [0x00BBBC97+518199]
	GetHandleVerifier [0x00BBAD80+514336]
	Ordinal0 [0x00ACED2D+2092333]
	Ordinal0 [0x00AD2EE8+2109160]
	Ordinal0 [0x00AD3022+2109474]
	Ordinal0 [0x00ADCB71+2149233]
	BaseThreadInitThunk [0x7561FA29+25]
	RtlGetAppContainerNamedObjectPath [0x76F47A9E+286]
	RtlGetAppContainerNamedObjectPath [0x76F47A6E+238]


In [ ]:
# 60 pages all scraped and save to csv first.

In [66]:
import re
temp = []

for i in range(len(global_reviews)):
    a = ''
    b = re.findall(r'[0-9\u4e00-\u9fff]+',global_reviews[i])
    temp.append([i, a.join(b)])

In [78]:
import csv

In [79]:
# 尝试import 到 csv
data = temp

with open("menshark_smurfs_pants_reviews_1603.csv", "w", newline="") as f:
    writer = csv.writer(f)
    writer.writerows(data)

In [82]:
len(temp)

1403

我们希望把关键词和对应的频率提取出来

In [109]:
from snownlp import SnowNLP

In [95]:
sentence = []
for x in temp:
    sentence.append(x[1])

In [145]:
words_cnt = {}
for x in set(sentence):
    if x == '':
        continue
    nlp_words = SnowNLP(x).words
    for y in nlp_words:
        if y in words_cnt:
            words_cnt[y] += 1
        else:
            words_cnt[y] = 1

In [146]:
# here words_cnt is a dict
def remove_wuxiao_char(words_cnt):
    words_cnt.pop('很')
    words_cnt.pop('的')
    words_cnt.pop('好')
    words_cnt.pop('了')
    words_cnt.pop('穿')
    words_cnt.pop('买')
    words_cnt.pop('也')
    words_cnt.pop('是')
    words_cnt.pop('挺')
    words_cnt.pop('还')
    words_cnt.pop('间')
    words_cnt.pop('就')
    words_cnt.pop('有')
    words_cnt.pop('一')
    words_cnt.pop('给')
    words_cnt.pop('抢')
    words_cnt.pop('比')
    words_cnt.pop('会')
    words_cnt.pop('我')
    words_cnt.pop('们')
    words_cnt.pop('在')
    words_cnt.pop('非常')
    words_cnt.pop('中')
    words_cnt.pop('对')
    words_cnt.pop('但是')
    words_cnt.pop('来')
    words_cnt.pop('点')
    words_cnt.pop('显')
    words_cnt.pop('个')
    words_cnt.pop('分')
    words_cnt.pop('天')
    return words_cnt

In [147]:
words_cnt = remove_wuxiao_char(words_cnt)

In [130]:
words_cnt_list = []
for x,y in words_cnt.items():
    words_cnt_list.append((x,y))

In [133]:
sorted_words_cnt_list = sorted(words_cnt_list, key = lambda x: x[1], reverse = True)
sorted_words_cnt_list[:150]

[('质量', 537),
 ('不错', 323),
 ('裤子', 315),
 ('厚实', 302),
 ('好看', 192),
 ('直播', 188),
 ('喜欢', 170),
 ('不', 163),
 ('娅', 151),
 ('价格', 134),
 ('舒服', 132),
 ('可以', 120),
 ('薇', 109),
 ('精灵', 108),
 ('满意', 107),
 ('蓝', 107),
 ('着', 105),
 ('面料', 95),
 ('合适', 92),
 ('大', 83),
 ('太', 82),
 ('推荐', 81),
 ('看', 78),
 ('划算', 69),
 ('没', 67),
 ('这个', 65),
 ('高', 63),
 ('可爱', 63),
 ('值得', 62),
 ('到', 61),
 ('节', 60),
 ('价', 59),
 ('实惠', 58),
 ('码', 55),
 ('便宜', 51),
 ('条', 50),
 ('收到', 50),
 ('上', 49),
 ('超级', 49),
 ('发货', 48),
 ('说', 48),
 ('图案', 47),
 ('都', 47),
 ('次', 47),
 ('快', 47),
 ('购买', 46),
 ('比较', 46),
 ('宝贝', 45),
 ('好评', 44),
 ('手', 44),
 ('老', 43),
 ('款式', 42),
 ('货', 41),
 ('型', 41),
 ('真的', 41),
 ('物流', 41),
 ('速度', 41),
 ('适合', 41),
 ('舒适', 40),
 ('特别', 40),
 ('性', 39),
 ('正好', 37),
 ('速', 37),
 ('宽松', 36),
 ('服', 36),
 ('没有', 36),
 ('做工', 36),
 ('包装', 36),
 ('版', 35),
 ('超值', 35),
 ('衣服', 34),
 ('小', 34),
 ('有点', 34),
 ('棉', 33),
 ('颜色', 33),
 ('吧', 32),
 ('黑色', 32),
 ('又', 32),
 

In [134]:
# 尝试import 到 csv
data = sorted_words_cnt_list[:150]

with open("menshark_smurfs_pants_reviews_top_150.csv", "w", newline="") as f:
    writer = csv.writer(f)
    writer.writerows(data)

感觉还是有很多词被拆开了，所以重新整合一下

In [148]:
# ’质量好‘ 和 ’质量差‘ ,薇娅，物流慢，物流快
zl_hao = 0
zl_cha = 0
wl_k = 0 
wl_m = 0
viya = 0
mao = 0
for x in set(sentence):
    if '质量' and '差' in x:
        zl_cha += 1
    if '质量' and '好' in x:
        zl_hao += 1
    if '薇娅' in x:
        viya += 1
    if '物流' and '慢' in x:
        wl_m += 1
    if '物流' and '快' in x:
        wl_k += 1
    if '毛' in x:
        mao += 1
    else:
        continue

In [ ]:
words_cnt

In [ ]:
words_cnt.pop('质量')

In [157]:
words_cnt['质量好'] = zl_hao
words_cnt['质量差'] = zl_cha
words_cnt['物流快'] = wl_k
words_cnt['物流慢'] = wl_m
words_cnt['薇娅'] = viya
words_cnt['掉毛'] = mao

In [160]:
words_cnt.pop('不')
words_cnt.pop('薇')
words_cnt.pop('娅')
words_cnt.pop('着')

56

In [161]:
words_cnt_list_revised = []
for x,y in words_cnt.items():
    words_cnt_list_revised.append((x,y))

In [162]:
sorted_words_cnt_list_revised = sorted(words_cnt_list_revised, key = lambda x: x[1], reverse = True)
sorted_words_cnt_list_revised[:150]

[('质量好', 372),
 ('裤子', 165),
 ('不错', 155),
 ('厚实', 150),
 ('好看', 88),
 ('直播', 85),
 ('喜欢', 85),
 ('薇娅', 70),
 ('舒服', 66),
 ('可以', 66),
 ('价格', 65),
 ('精灵', 58),
 ('蓝', 57),
 ('面料', 50),
 ('合适', 48),
 ('大', 43),
 ('满意', 43),
 ('太', 40),
 ('看', 37),
 ('这个', 37),
 ('推荐', 36),
 ('没', 36),
 ('物流快', 35),
 ('到', 33),
 ('划算', 32),
 ('可爱', 32),
 ('节', 30),
 ('高', 30),
 ('条', 29),
 ('价', 29),
 ('便宜', 29),
 ('实惠', 28),
 ('图案', 26),
 ('说', 25),
 ('上', 24),
 ('老', 24),
 ('都', 24),
 ('次', 23),
 ('适合', 23),
 ('值得', 23),
 ('码', 23),
 ('收到', 23),
 ('比较', 23),
 ('发货', 23),
 ('手', 22),
 ('款式', 22),
 ('宝贝', 22),
 ('快', 22),
 ('性', 21),
 ('货', 21),
 ('速', 20),
 ('服', 20),
 ('特别', 20),
 ('好评', 20),
 ('舒适', 19),
 ('购买', 19),
 ('真的', 19),
 ('正好', 19),
 ('物流', 19),
 ('型', 18),
 ('宽松', 18),
 ('没有', 18),
 ('有点', 18),
 ('超级', 18),
 ('小', 18),
 ('真', 17),
 ('又', 17),
 ('感觉', 17),
 ('棉', 17),
 ('做工', 16),
 ('想', 16),
 ('孩子', 16),
 ('态度', 16),
 ('超值', 16),
 ('吧', 16),
 ('黑色', 16),
 ('毛', 16),
 ('速度', 16),
 ('这', 15)

In [165]:
# 尝试import 到 csv
data = sorted_words_cnt_list_revised[:150]

with open("menshark_smurfs_pants_reviews_top_150.csv", "w", newline="") as f:
    writer = csv.writer(f)
    writer.writerows(data)

In [167]:
len(set(sentence))

673

In [169]:
# 提到掉毛的频率
13 / 673

0.019316493313521546

In [ ]:
# 接下来我们来特别分析一下差评


In [177]:
kefu = []
for x in set(sentence):
    if '客服' in x:
        kefu.append(x)

In [184]:
import re

In [205]:
re.findall('\客\服..', kefu[9])

['客服小乔']

In [208]:
kefu_names = []
for x in kefu:
    kefu_names.append(re.findall('\客\服..', x))

In [211]:
kefu_dict = {}
kefu_dict['客服态度不好'] = 7
kefu_dict['客服态度好'] = 12
kefu_dict['客服小乔态度耐心'] = 2
kefu_dict['客服小周态度好'] = 1
kefu_dict['客服小柚态度负责'] = 1
kefu_dict['客服小叶态度贴心'] = 1
kefu_dict['客服小豆返现不理人'] = 1

In [213]:
kefu_list = []
for x,y in kefu_dict.items():
    kefu_list.append([x,y])

In [216]:
data = kefu_list
with open("menshark_smurfs_pants_reviews_kefu.csv", "w", newline="") as f:
    writer = csv.writer(f)
    writer.writerows(data)

In [221]:
len(set(sentence))

673

In [219]:
len(kefu)

19

蓝精灵卫衣一共543条有效评论中提到关键词‘质量好’的评论有151条，占比为28%；值得注意的是提到关键词‘掉毛’的评论有71条，占比高达13%.

蓝精灵裤子一共673条有效评论中提到关键词‘质量好’的评论有372条，占比达到 55.3%；值得注意的是提到关键词‘掉毛’的评论只有13条，占比约为 2%，相较于蓝精灵卫衣低了11个百分点。673条有效评论中提到‘客服’的有19条，其中表扬客服有12条，7条是吐槽客服态度不好的，值得注意的是客服小乔被特别表扬2次，总表扬数中占比为16%.